tests/test_core.py

```diff
@@ -55,6 +56,30 @@ def test_ones_like():
   t2j_function_test(torch.ones_like, [(2, 3)], tests=tests)
 
 
+def test_tree_coerce():
+  @dataclass
+  class A:
+    a: torch.Tensor
+
+  @dataclass
+  class B:
+    a: A
+    b: int
+
+  torch.utils._pytree.register_dataclass(A)
+  torch.utils._pytree.register_dataclass(B)
+
+  def f():
+    t1 = torch.return_types.max([1, torch.ones(3)])
+    t2 = torch.return_types.topk([3.0, t1])
+    t3 = A(torch.ones(3, 2))
+    t4 = B(t3, 7)
+    return (t2, t3, t4)
+
+  tests = [forward_test]
+  t2j_function_test(f, [], tests=tests)
```

Comment:
I'm worried that this test is missing a case: `forward_test` always calls the non-`jit` version of `f` before the `jit` version of `f`. that means we're not currently testing the case where we are calling `jit(f)` first. I'm concerned that `register_pytree_node` may not work in the jit context, and that we're just getting lucky bc we're calling the non-jit `f` first which does the registration for us. 

let's add a test that only calls the jit version to ensure that it works that way as well

Reply:
Test added, confirms it works with jit. Jit will always run the python code first for tracing, and the registration part is pure python, which jax will execute as it is.